### ISSUE TITLE
Autogenerated `__init__` is not picked up when using a decorator that wraps `@dataclass` 

### ISSUE THREAD

**Issue by j1ah0ng** [microsoft/pyright#7062]
**Describe the bug**

When wrapping a stdlib `dataclass` decorator to create a new decorator, Pyright doesn't appear to pick up the autogenerated `__init__` implementation.

**Code or Screenshots**

With the following snippet:
```python
# test.py
from dataclasses import dataclass

def _d(*args, **kwargs):
    return lambda cls: dataclass(*args, frozen=True, **kwargs)(cls)

@_d
class testtt:
    a: bool

if __name__ == "__main__":
    testtt(True)
```
the REPL works fine:
```python
>>> from dataclasses import dataclass
>>> def _d(*args, **kwargs):
...     return lambda cls: dataclass(*args, frozen=True, **kwargs)(cls)
>>> @_d
... class testtt:
...     a: bool
>>> testtt(False)
testtt(a=False)
```
but Pyright complains about the argument to `testtt`'s init:
```sh
❯ pyright test.py
test.py
  test.py:11:5 - error: Expected no arguments to "testtt" constructor (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```
This behaviour is mirrored in `coc-nvim`:
<img width="763" alt="image" src="https://github.com/microsoft/pyright/assets/13944989/c1129424-e55d-4a96-827d-6bde765b2cc2">

**VS Code extension or command-line**

`coc-nvim` output:
```
2024-01-20T13:07:48.951 INFO (pid:26623) [plugin] - coc.nvim initialized with node: v21.5.0 after 181
2024-01-20T13:07:48.952 INFO (pid:26623) [services] - LanguageClient Pyright Server state change: stopped => starting
2024-01-20T13:07:48.955 INFO (pid:26623) [language-client-index] - Language server "pyright" started with 26692
2024-01-20T13:07:49.095 INFO (pid:26623) [services] - LanguageClient Pyright Server state change: starting => running
```
command line:
```sh
❯ pyright --version
pyright 1.1.347
```

**Comment by erictraut** [comment]
This behavior is "as designed". It's not a bug. The `dataclass` decorator requires significant special-case handling in a type checker, and the arguments passed to `dataclass` (like the values for the `init` or `frozen` parameters) must be statically evaluated. The way you're wrapping `dataclass` in your code sample cannot be handled by a static type analyzer.

As a workaround, you could make use of `dataclass_transform` (introduced in [PEP 681](https://peps.python.org/pep-0681/)) to indicate that your wrapper applies dataclass-like semantics.

**Comment by j1ah0ng** [comment]
Ack, thanks for the pointer.
